[MLIR] Generalize expand_shape to take shape as explicit input

Shukla-Gaurav opened the pull request:
This patch generalizes tensor.expand_shape and memref.expand_shape to consume the output shape as a list of SSA values.  This enables us to implement generic reshape operations with dynamic shapes using collapse_shape/expand_shape pairs.

The output_shape input to expand_shape follows the static/dynamic representation that's also used in `tensor.extract_slice`.

Differential Revision: https://reviews.llvm.org/D140821

ramiro050 committed:
[MLIR] Generalize expand_shape to take shape as explicit input

This patch generalizes tensor.expand_shape and memref.expand_shape to consume
the output shape as a list of SSA values.  This enables us to implement generic
reshape operations with dynamic shapes using collapse_shape/expand_shape pairs.

The output_shape input to expand_shape follows the static/dynamic representation
that's also used in `tensor.extract_slice`.

Differential Revision: https://reviews.llvm.org/D140821

MaheshRavishankar commented on mlir/include/mlir/Dialect/Utils/StaticValueUtils.h:
```diff
@@ -109,9 +109,8 @@ SmallVector<OpFoldResult> getMixedValues(ArrayRef<int64_t> staticValues,
-std::pair<ArrayAttr, SmallVector<Value>>
-decomposeMixedValues(Builder &b,
-                     const SmallVectorImpl<OpFoldResult> &mixedValues);
+std::pair<SmallVector<int64_t>, SmallVector<Value>>
```
I am not sure this is needed.

Shukla-Gaurav replied:
Yeah, it seems this patch does not necessarily require this method. If the helper build method seems fine(mentioned in the next comment), I will address this comment together with the next one.

MaheshRavishankar commented on mlir/lib/Conversion/TosaToTensor/TosaToTensor.cpp:
```diff
@@ -197,8 +197,21 @@ class ReshapeConverterExpand : public OpConversionPattern<tosa::ReshapeOp> {
+    SmallVector<OpFoldResult> inputShape =
+        tensor::getMixedSizes(rewriter, reshape.getLoc(), input);
+    SmallVector<OpFoldResult> outputShape;
+    if (failed(tensor::ExpandShapeOp::inferOutputShape(
```
We could just use a helper build method to do this automatically instead of adding it at every call site.

Shukla-Gaurav replied:
I have added a helper build method which can infer the output shape. But in the cases where we can not infer output shapes, it's assert vs failure. With the helper build method, in case the output shapes can not be inferred, it breaks with assert(let me know if there is better way?). 

If the helper build method seems fine, I will update other parts of the code to use this build method.
cc: @ramiro050

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -454,9 +454,17 @@ struct FoldFillWithTensorReshape : OpRewritePattern<TensorReshapeOp> {
-        loc, reshapeOp.getResultType(), oldFill.output(),
-        reshapeOp.getReassociation());
+    TensorReshapeOp newInit;
+    if constexpr (std::is_same<TensorReshapeOp, tensor::ExpandShapeOp>::value)
```
Nit: add `{` `}` for multiline statements.

MaheshRavishankar commented on mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorRewriting.cpp:
```diff
@@ -453,8 +453,15 @@ struct ReshapeRewriter : public OpRewritePattern<ReshapeOp> {
-      auto reshape = rewriter.create<ReshapeOp>(loc, denseTp, op.getSrc(),
-                                                op.getReassociation());
+      ReshapeOp reshape;
+      if constexpr (std::is_same<ReshapeOp, tensor::ExpandShapeOp>::value) {
```
Same here. Use `{` `}` for multi-line statements.

Shukla-Gaurav replied:
Did you mean some other if stmt, as this one already have.

MaheshRavishankar commented on mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:
```diff
@@ -16,6 +17,76 @@
+      if (ShapedType::isDynamic(expandedType.getDimSize(index))) {
+        // Call get<Value>() under the assumption that we're not casting
+        // dynamism.
+        Value indexGroupSize = inputShape[inputIndex].get<Value>();
```
We want to avoid creating `arith.divui` ops directly. Use `affine.apply` operations instead. Those fold better.

Shukla-Gaurav replied:
working on this.

Shukla-Gaurav replied:
Thanks, updated with affine.apply.

MaheshRavishankar reviewed (changes requested):
Ok, I think this is not looking too bad. Adding the helper build method here would make the diff much smaller I think.

MaheshRavishankar commented:
Hey, I left a few comments. Let me know when this is ready for review.

Shukla-Gaurav commented on mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h:
```diff
@@ -167,9 +190,11 @@ static LogicalResult verifyReshapeLikeShapes(OpTy op, ShapedType collapsedType,
 /// Pattern to collapse producer/consumer reshape ops that are both collapsing
 /// dimensions or are both expanding dimensions.
-template <typename ReshapeOpTy>
+template <typename ReshapeOpTy, ReshapeOpKind opKind>
```
Yeah, we can get it by ReshapeOpTy itself.
I guess this extra template parameter maybe added to simplify the check as this utility can be used by reshape ops of many dialects(currently used by memref and tensor dialects). @ramiro050 may have more context on this.

ramiro050 replied:
See: https://reviews.llvm.org/D140821#inline-1369788

ramiro050 replied:
From Sanjoy's comment in link above:

> Unless I'm missing some C++ trick, that requires to add a dependency from this header to either the memref or the tensor dialect. I assumed that was not OK, but please let me know if it is.

Shukla-Gaurav commented on mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h:
```diff
@@ -30,6 +30,29 @@ using ReassociationExprs = SmallVector<AffineExpr, 2>;
+// Returns the output shape in `outputShape` and `staticOutputShape`, following
+// the conventions for the output_shape and static_output_shape inputs to the
+// expand_shape ops.
+LogicalResult
```
This method only returns success or failure, the output_shape is pass by reference.

MaheshRavishankar commented on mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:
```diff
@@ -1050,39 +1050,83 @@ def Tensor_ExpandShapeOp : Tensor_ReassociativeReshapeOp<"expand_shape"> {
+    [{
+      SmallVector<OpFoldResult> inputShape = getMixedSizes($_builder, $_state, src);
+      SmallVector<OpFoldResult> outputShape;
+      assert(failed(inferOutputShape($_builder, $_state, resultType, reassociation, 
```
Please take these out of `asserts` . This will be ignored during release builds. SO
```
auto status = inferOutputShape(...)
(void) status
assert(succeeded(status) && ...);
```

MaheshRavishankar commented on mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp:
```diff
@@ -550,9 +549,20 @@ static Value expandRank(PatternRewriter &rewriter, Location loc, Value tensor,
+  SmallVector<OpFoldResult> inputShape =
+      tensor::getMixedSizes(rewriter, loc, tensor);
+  SmallVector<OpFoldResult> outputShape;
+  if (failed(tensor::ExpandShapeOp::inferOutputShape(
```
With the new build method you should be able to fold this into the `create` method? Here and everywhere else?

Shukla-Gaurav replied:
Updated at all places, thanks!

MaheshRavishankar commented on mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:
```diff
@@ -16,6 +17,76 @@
+    }
+  }
+
+  assert(static_cast<uint64_t>(
```
Just return `failure` here.

Also it seems like there is really no `failure` returning happening in this method. One option is to make this function return `void` (in which case the `assert` here is OK). On balance, maybe better to just keep the `LogicalResult` return type and return `failure()`.

MaheshRavishankar reviewed (changes requested):
Next rounds of comments. Thanks!

Shukla-Gaurav commented:
@MaheshRavishankar Thanks for reviewing, have addressed all the comments, and is ready for re-review.
@ramiro050 @sanjoy Thanks for implementing this, it was needed for a long time. As I am continuing this, would love to have your review on this.

MaheshRavishankar reviewed (commented):
ok, this looks good to me. Lets test this with IREE. Ill let you know how to do that.

ramiro050 reviewed (approved):
Thanks for getting this across the finish line! LGTM

newling commented on mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:
```diff
@@ -16,6 +18,83 @@
+      (outputShapeValues.size() - llvm::count(outputShapeValues, Value{})))
+    return failure();
+
+  llvm::erase(outputShapeValues, Value{});
```
nit:

outputShapeValues has the same size as the input's rank. I think it could be initialized at this size, and populated by the reassociation group index. i.e. 

 `outputShapeValues[inputIndex] = dynamicDimSize;` instead of `outputShapeValues[index] = dynamicDimSize;`

Shukla-Gaurav replied:
Thanks for pointing this, updated the code :)

newling commented on mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:
```diff
@@ -230,24 +309,17 @@ LogicalResult mlir::reshapeLikeShapesAreCompatible(
-      if (ShapedType::isDynamic(dim.value())) {
-        if (isExpandingReshape && dynamicShape) {
-          return emitError("invalid to have a single dimension (" +
-                           Twine(map.index()) +
```
aI'd like to get a better understanding of this PR from a high-level. Is it now permitted to have multiple dynamic dimensions in an association? 

Currently this is what happens when you try to expand to multiple dynamic dimensions:

```
foo.mlir:3:8: error: 'tensor.expand_shape' op invalid to have a single dimension (0) expanded into multiple dynamic dims (0,1)
  %0 = tensor.expand_shape %arg0 [[0, 1], [2]] : tensor<?x?xf32> into tensor<?x?x?xf32>
  ```
  
 Is the idea that with this PR this is allowed, but only as long as the dynamic dimensions have values associated to them?
  
  I guess one high-level question is, does this make expand_shape more general purpose, and does that mean that the optimisations that can be applied to it are less powerful?
  
  Any additional high-level info would be welcome :-)

MaheshRavishankar replied:
I think this is generally increasing the expressive power of the expand_shape op. You are right, it allows you to represent "how" a dynamic dimension is split into multiple dynamic dimensions. Most transformations AFAIK should not care about this split. The load-bearing information is know how the dimensions are split, i.e. the information being carried by reassociation indices... 
I wouldnt say its making expand_shape more general purpose. You are still splitting a dimension, into multiple, its just that those multiple dimensions can now be dynamic. The information about the final output shape is now carried in the op, making the split deterministic. Not sure if that makes sense, but I can explain on a call if that help.s

Shukla-Gaurav commented:
Currently debugging following issue in this PR:(this issue arises while testing the patch in IREE https://github.com/openxla/iree/pull/16972 )

**add_static.mlir**
```
func.func @main(%input : tensor<1x8x8x3xf32>) -> (tensor<1x8x8x3xf32>) {
  %1 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>, affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>, affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%input, %input : tensor<1x8x8x3xf32>, tensor<1x8x8x3xf32>) outs(%input : tensor<1x8x8x3xf32>) {
  ^bb0(%in: f32, %in_0: f32, %out: f32):
    %3 = arith.addf %in, %in_0 : f32
    linalg.yield %3 : f32
  } -> tensor<1x8x8x3xf32>
  return %1 : tensor<1x8x8x3xf32>
}
```
`/bin/mlir-opt  --test-linalg-drop-unit-dims add_static.mlir`
[stack_dump.txt](https://github.com/llvm/llvm-project/files/14874018/stack_dump.txt)

cc: @MaheshRavishankar @pashu123

MaheshRavishankar commented:
> Currently debugging following issue in this PR:(this issue arises while testing the patch in IREE [openxla/iree#16972](https://github.com/openxla/iree/pull/16972) )
> 
> **add_static.mlir**
> 
> ```
> func.func @main(%input : tensor<1x8x8x3xf32>) -> (tensor<1x8x8x3xf32>) {
>   %1 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>, affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>, affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%input, %input : tensor<1x8x8x3xf32>, tensor<1x8x8x3xf32>) outs(%input : tensor<1x8x8x3xf32>) {
>   ^bb0(%in: f32, %in_0: f32, %out: f32):
>     %3 = arith.addf %in, %in_0 : f32
>     linalg.yield %3 : f32
>   } -> tensor<1x8x8x3xf32>
>   return %1 : tensor<1x8x8x3xf32>
> }
> ```
> 
> **Command** `/bin/mlir-opt --test-linalg-drop-unit-dims add_static.mlir` [stack_dump.txt](https://github.com/llvm/llvm-project/files/14874018/stack_dump.txt)
> 
> cc: @MaheshRavishankar @pashu123

Looking at the stack error seems like it isnt calling the "right" version of the build....

aartbik commented on mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorRewriting.cpp:
```diff
@@ -952,8 +952,15 @@ struct ReshapeRewriter : public OpRewritePattern<ReshapeOp> {
-      auto reshape = rewriter.create<ReshapeOp>(loc, denseTp, op.getSrc(),
-                                                op.getReassociation());
+      ReshapeOp reshape;
+      if constexpr (std::is_same<ReshapeOp, tensor::ExpandShapeOp>::value) {
```
can this be hidden inside a templated "create" helper method, so we don't get such a huge if-then-else block?

Shukla-Gaurav replied:
There are issues with templated helper. There are few more things which has to be handled for this patch like fusion, I can raise a PR with this and other issues fixed. It has lots of changes, so trying to not hold it for longer. Thanks :)

bjacob replied:
+1, i have a bunch of things I want to do that depend on this, thanks for deferring improvements to follow-up PRs!

I'm also available to help with C++ templates if needed.

ScottTodd commented on mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:
```diff
@@ -8,6 +8,7 @@
 
 #include "mlir/Dialect/Utils/ReshapeOpsUtils.h"
 
+#include "mlir/Dialect/Arith/IR/Arith.h"
 #include "mlir/IR/AffineMap.h"
 #include "mlir/IR/Builders.h"
```
Is this dependency from `Dialect/Utils/` onto the Arith dialect acceptable, or should this new `inferExpandShapeOutputShape()` function be moved to somewhere else (`Dialect/Arith/Utils`? Or memref _and_ tensor where the code is used, with some code duplication?)

Shukla-Gaurav replied:
I tried moving it to tensor/utils, but there seem to be circular dependencies. I can handle this in the next PR. Can we go ahead for this patch as it is?

bjacob replied:
+1, see my other comment, thanks for deferring further improvements to follow-up PRs!

Shukla-Gaurav commented:
@MaheshRavishankar Could you please re-review it? There is still one failure in reshape_fusion(as discussed), meanwhile I can fix that. Thanks :)

aartbik commented:
I added a comment that has not been addressed yet?

MaheshRavishankar commented on mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:
```diff
@@ -1102,50 +1097,103 @@ def Tensor_ExpandShapeOp : Tensor_ReassociativeReshapeOp<"expand_shape"> {
+    OpBuilder<(ins "Type":$resultType, "Value":$src,
+      "ArrayRef<ReassociationIndices>":$reassociation),
+    [{
+      SmallVector<OpFoldResult> inputShape = 
```
Can we move the implementation of this into c++ file `Tensor/IR/TensorOps.cpp`. Not good practice to have implementation in the tablegen file, especially when its doing non-trivial things.

Shukla-Gaurav replied:
Moved the implementation to TensorOps.cpp for the first two builders as they contain non-trivial code. Thanks :)

MaheshRavishankar commented on mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:
```diff
@@ -1102,50 +1097,103 @@ def Tensor_ExpandShapeOp : Tensor_ReassociativeReshapeOp<"expand_shape"> {
-      build($_builder, $_state, resultType, src, attrs);
-      $_state.addAttribute("reassociation",
-          getReassociationIndicesAttribute($_builder, reassociation));
+      auto [staticOutputShape, dynamicOutputShape] =
```
While here, maybe move these to c++ file as well.

MaheshRavishankar commented on mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:
```diff
@@ -1102,50 +1097,103 @@ def Tensor_ExpandShapeOp : Tensor_ReassociativeReshapeOp<"expand_shape"> {
+      SmallVector<OpFoldResult> inputShape = 
+          getMixedSizes($_builder, $_state.location, src);
+      std::pair<SmallVector<int64_t>, SmallVector<Value>> outputShape;
+      auto status = 
```
total nit: but if you want to do this, the canonical way of doing this would be to return
```
FailureOr<SmallVector<OpFoldResult>> inferOutputShape(...)
```

That way you can return a mix of values and int64_t as a single vector and not have to return by reference.

MaheshRavishankar commented on mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:
```diff
@@ -1622,29 +1613,72 @@ def MemRef_ExpandShapeOp : MemRef_ReassociativeReshapeOp<"expand_shape", [
+    OpBuilder<(ins "Type":$resultType, "Value":$src,
+      "ArrayRef<ReassociationIndices>":$reassociation),
+    [{
+      SmallVector<OpFoldResult> inputShape = 
```
I saw this file after the `TensorOps.td` file. So same comments from there apply here too.. Move these to C++ file

MaheshRavishankar commented on mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:
```diff
@@ -1622,29 +1613,72 @@ def MemRef_ExpandShapeOp : MemRef_ReassociativeReshapeOp<"expand_shape", [
+      SmallVector<OpFoldResult> inputShape = 
+          getMixedSizes($_builder, $_state.location, src);
+      std::pair<SmallVector<int64_t>, SmallVector<Value>> outputShape;
+      auto status = 
```
Nit : Echoing comment from earlier. This would be much more readable with `FailureOr<SmallVector<OpFoldResul>>` . Fine to do it later too.

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -586,9 +586,18 @@ struct FoldFillWithTensorReshape : OpRewritePattern<TensorReshapeOp> {
-        reshapeOp.getReassociation());
+    TensorReshapeOp newInit;
+    if constexpr (std::is_same<TensorReshapeOp, tensor::ExpandShapeOp>::value) {
+
```
nit: unnecessary empty line

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp:
```diff
@@ -757,7 +757,10 @@ pushDownUnPackOpThroughExpandShape(tensor::UnPackOp unPackOp,
   ArrayRef<int64_t> outerDimsPerm = unPackOp.getOuterDimsPerm();
 
-  ArrayRef<int64_t> dstShape = expandOp.getType().getShape();
+  auto expandTy = expandOp.getType().dyn_cast<RankedTensorType>();
```
Why did you have to change this? This should have always been a ranked tensor type. If so then something is off with the pack op. Fine to leave as is, just noticing.

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/Transforms/DropUnitDims.cpp:
```diff
@@ -252,7 +253,7 @@ replaceUnitDimIndexOps(GenericOp genericOp,
 /// The `reassociation` is used when `rankReductionStrategy` is set to
 /// `RankReductionStrategy::ReassociativeReshape`.
-static Value
+static FailureOr<Value>
```
This is not returning a failure yet? So then this doesnt need to be a `FailureOr`

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp:
```diff
@@ -815,16 +817,28 @@ fuseWithReshapeByExpansion(LinalgOp linalgOp, Operation *reshapeOp,
                 /*isExpandingReshape=*/true)))
           return std::nullopt;
+
+        SmallVector<OpFoldResult> inputShape =
```
If you change the `isFusableMethod` you can just remove this part I think.

MaheshRavishankar commented on mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h:
```diff
@@ -30,6 +30,28 @@ using ReassociationExprs = SmallVector<AffineExpr, 2>;
+// the conventions for the output_shape and static_output_shape inputs to the
+// expand_shape ops.
+LogicalResult inferExpandShapeOutputShape(
+    OpBuilder &b, Location loc, RankedTensorType expandedType,
```
I see later that you are using the same method for `MemRefType` and `RankedTensorType`. Maybe make the `expandedType` be `ShapedType` that way you dont need to create a tensor type just to call this method.

MaheshRavishankar commented on mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:
```diff
@@ -2237,6 +2237,17 @@ FailureOr<MemRefType> ExpandShapeOp::computeExpandedType(
+    ArrayRef<ReassociationIndices> reassociation,
+    ArrayRef<OpFoldResult> inputShape,
+    std::pair<SmallVector<int64_t>, SmallVector<Value>> &outputShape) {
+  auto expandedTensorType =
```
Note: This triggered the suggestion of using `ShapedType` in `inferExpandShapeOutputShape` to avoid having to do this.

MaheshRavishankar commented on mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:
```diff
@@ -230,24 +294,17 @@ LogicalResult mlir::reshapeLikeShapesAreCompatible(
   unsigned expandedDimStart = 0;
   for (const auto &map : llvm::enumerate(reassociationMaps)) {
-    std::optional<int64_t> dynamicShape;
+    bool foundDynamicShape = false;
```
Nice cleanup!

Shukla-Gaurav replied:
Thanks :)

MaheshRavishankar reviewed (changes requested):
Thanks! Left a few mostly minor comments. Nothing that I see as blocker for landing.

bjacob commented on mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:
````diff
@@ -1598,14 +1593,10 @@ def MemRef_ExpandShapeOp : MemRef_ReassociativeReshapeOp<"expand_shape", [
     ```mlir
-    %r = memref.expand_shape %0 [[0, 1], [2]]
-        : memref<?x?xf32> into memref<?x5x?xf32>
+    %r = memref.expand_shape %0 [[0, 1], [2]] [%sz0, %sz1, 32]
````
I think this example is missing `output_shape` ?

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp:
```diff
@@ -778,6 +805,11 @@ fuseWithReshapeByExpansion(LinalgOp linalgOp, Operation *reshapeOp,
 
+  // TODO: With the support of multiple dynamic dims expansion in
+  // tensor.expand_shape op, this case can be handled.
+  if (failed(validateDynamicDimExpansion(linalgOp, expansionInfo, rewriter)))
```
Can we move this check into `isFusableWithReshapeByDimExpansion`?

Shukla-Gaurav replied:
Yes, this needs to be fixed anyway and it was easier to place this check here as we have expansionInfo.

MaheshRavishankar reviewed (approved):
Ok, looks ogod to me. Left a single comment but it is not a blocker and is OK to do it after this PR as well since that part needs to be fixed after this PR lands.

bjacob commented:
> Ok, looks ogod to me. Left a single comment but it is not a blocker and is OK to do it after this PR as well since that part needs to be fixed after this PR lands.

yes please please do as follow-up and not lose the progress on (very slow) CI here :-)

bjacob commented:
Windows CI failed with

> fatal error C1060: compiler is out of heap space

Probably unrelated to this PR, but also not completely impossible that something about this PR increases compiler memory usage and pushes it over a threshold, so we need to get a green CI run before merging. Unfortunately I don't see a "retrigger" button here. Can you push an empty commit to this PR to cause a retrigger maybe?

joker-eph commented:
I think this broke the bot: https://lab.llvm.org/buildbot/#/builders/61/builds/57064

joker-eph commented:
Reverted in https://github.com/llvm/llvm-project/pull/89540
